### ISSUE TITLE
How to use pickneglogsoftmax()?

### ISSUE THREAD

**Issue by harumo11** [clab/dynet#1571]
Hello, developer of Dynet and everyone.

I can't understand how to use `pickneglogsoftmax()`.

In order to solve [iris](https://medium.com/@jebaseelanravi96/machine-learning-iris-classification-33aa18a4a983) problem, I tried to write following code.
But, this code caused **segmentation fault**.

```C++
	// Set parameters
	// W1 (8x4) matrix
	dynet::Parameter p_W1 = model.add_parameters({HIDDEN_LAYER_SIZE, INPUT_LAYER_SIZE});
	// b1 (16x1) vector
	dynet::Parameter p_b1 = model.add_parameters({HIDDEN_LAYER_SIZE});
	// W2 (3x16) matrix
	dynet::Parameter p_W2 = model.add_parameters({OUTPUT_LAYER_SIZE, HIDDEN_LAYER_SIZE});
	// b2 (3x1) vector
	dynet::Parameter p_b2 = model.add_parameters({OUTPUT_LAYER_SIZE});

	// Making nodes and registering to computational graph
	dynet::ComputationGraph cg;
	dynet::Expression W1 = dynet::parameter(cg, p_W1);
	dynet::Expression b1 = dynet::parameter(cg, p_b1);
	dynet::Expression W2 = dynet::parameter(cg, p_W2);
	dynet::Expression b2 = dynet::parameter(cg, p_b2);

	// Making input variable x_value
	std::vector<dynet::real> x_value(4);
	dynet::Expression x = dynet::input(cg, {4}, x_value);

	// Defintions of relationship of nodes
	dynet::Expression z1 = dynet::rectify(W1*x+b1);		// Input layer
	dynet::Expression y_pred = dynet::softmax(W2*z1+b2);// Hidden layer

	// Making output value y_label
	unsigned* y_label = 0;

	dynet::Expression loss_expr = dynet::pickneglogsoftmax(y_pred, y_label);
```

As you can see, I used `pickneglogsoftmax(const Expression&, unsinged* pv)` which is a pointer version at this time.

At next time, I followed a [advice](https://groups.google.com/forum/#!topic/dynet-users/lAxqbIyhPj0) from @neubig and chenged loss function `pickneglogsoftmax(const Expression& x, unsinged v)` .

And I build computational graph as below.

![graph](https://user-images.githubusercontent.com/7510966/60857916-4ecdd200-a247-11e9-8cb9-bbd0a3aff939.png)

But, error was not converged with this computational graph as below.

![error](https://user-images.githubusercontent.com/7510966/60858004-a66c3d80-a247-11e9-8309-a837ce9e2a8b.png)

Please tell me how to use `pickneglogsoftmax(const Expression& x, unsinged* pv)` .

There is a whole source code [here](https://github.com/harumo11/dynet_start/blob/master/iris/iris.cpp).

If you would like to get further information, please feel free to contact me.
Thank you.

**Comment by shuheik** [comment]
It seems that you are calling pickneglogsoftmax on already softmax-ed expression y_pred. 
Try removing dynet::softmax around W2*z1+b2,

**Comment by harumo11** [comment]
Thanks for replying @shuheik !
And sorry for may late reply.

I removed `dynet::softmax` around W2*z1+b2.
But the loss did not converge.

![image](https://user-images.githubusercontent.com/7510966/61134847-73ea6b00-a4fb-11e9-94f1-900e8c231c00.png)

And the following computational graph was built.

![graph](https://user-images.githubusercontent.com/7510966/61134904-8fee0c80-a4fb-11e9-8aee-c33833cbe56a.png)

Is the following source code correct?

```
		// Setting concrete x_value
		x_value[0] = iris_data_table.data_table[itr].sepal_length;
		x_value[1] = iris_data_table.data_table[itr].sepal_width;
		x_value[2] = iris_data_table.data_table[itr].petal_length;
		x_value[4] = iris_data_table.data_table[itr].petal_width;

		// setosa	  
		if (iris_data_table.data_table[itr].class_name == "Iris-setosa"){
			*y_label = 0;
			std::cout << "[label] : " << *y_label << std::endl;
			loss_value = dynet::as_scalar(cg.forward(loss_expr));
			cg.backward(loss_expr);
		}
		// versicolor 
		if (iris_data_table.data_table[itr].class_name == "Iris-versicolor") {
			*y_label = 1;
			std::cout << "[label] : " << *y_label << std::endl;
			loss_value = dynet::as_scalar(cg.forward(loss_expr));
			cg.backward(loss_expr);
		}
		// virginica  
		if (iris_data_table.data_table[itr].class_name == "Iris-virginica") {
			*y_label = 2;
			std::cout << "[label] : " << *y_label << std::endl;
			loss_value = dynet::as_scalar(cg.forward(loss_expr));
			cg.backward(loss_expr);
		}
```

If you have a simple example, please take a look it.

**Comment by shuheik** [comment]
You are plotting losses of each example in one epoch. 
If you train the model for more epochs and calculate the average loss in each epoch, it will tend to decrease for a while. 
You may also need to adjust the learning rate to make it converge.
